src/processors/transform.rs

```diff
@@ -0,0 +1,293 @@
+        let mut matrix: Matrix4<f32> = (&quat * Matrix3::new(self.scale[0], 0.0, 0.0,
+                                                             0.0, self.scale[1], 0.0,
+                                                             0.0, 0.0, self.scale[2])).into();
+        matrix.w = Vector3::from(self.pos).extend(1.0f32);
+        matrix.into()
+    }
+}
+
+impl Default for LocalTransform {
+    fn default() -> Self {
+        LocalTransform {
+            pos: [0.0, 0.0, 0.0],
+            rot: [0.0, 0.0, 0.0, 1.0],
+            scale: [1.0, 1.0, 1.0],
+            parent: None,
+            dirty: true,
+        }
+    }
+}
+
+impl Component for LocalTransform {
+    type Storage = VecStorage<LocalTransform>;
+}
+
+/// Absolute transformation (transformed from origin), should be used for rendering position and orientation.
+#[derive(Debug, Copy, Clone)]
+pub struct Transform(pub [[f32; 4]; 4]);
+impl Transform {
+    pub fn identity() -> Self {
+        Transform(
+            [
+                [1.0, 0.0, 0.0, 0.0],
+                [0.0, 1.0, 0.0, 0.0],
+                [0.0, 0.0, 1.0, 0.0],
+                [0.0, 0.0, 0.0, 1.0],
+            ]
+        )
+    }
+}
+
+impl Component for Transform {
+    type Storage = VecStorage<Transform>;
+}
+
+/// Initialization component, added to entity with a `LocalTransform` component after the first update.
+#[derive(Default, Copy, Clone)]
+pub struct Init;
+impl Component for Init {
+    type Storage = NullStorage<Init>;
+}
+
+/// Transformation processor.
+/// Handles updating `Transform` components based on the `LocalTransform` component and parents.
+pub struct TransformProcessor {
+    indices: HashMap<Entity, usize>,
+    sorted: VecDeque<(Entity, Option<Entity>)>,
+}
+
+impl TransformProcessor {
+    pub fn new() -> TransformProcessor {
+        TransformProcessor {
+            indices: HashMap::new(),
+            sorted: VecDeque::new(),
+        }
+    }
+}
+
+impl Processor<Arc<Mutex<Context>>> for TransformProcessor {
+    fn run(&mut self, arg: RunArg, _: Arc<Mutex<Context>>) {
+        let (entities, mut locals, mut globals, mut init) = arg.fetch(|w| {
+            (w.entities(), w.write::<LocalTransform>(), w.write::<Transform>(), w.write::<Init>())
+        });
+
+        let mut new: Vec<Entity> = Vec::new();
+        for (entity, local, _) in (&entities, &locals, !&init).iter() {
+            self.indices.insert(entity, self.sorted.len());
+            self.sorted.push_back((entity, local.parent));
+            new.push(entity.clone());
+        }
+
+        for entity in new {
+            init.insert(entity, Init);
+        }
+
+        // Compute transforms (global) from local transforms and parents.
+        let mut swap_candidates: Vec<(Entity, Entity)> = Vec::new();
+        for &(entity, parent) in self.sorted.iter() {
+            let mut local = locals.get_mut(entity).unwrap();
+
+            if local.dirty {

```

Comment:
Hmm, this is actually a bit difficult of a fix. Since we could just use the parent's `LocalTransform` as a check if its dirty, but then any changes made by other systems to the global transform are ignored.